File: android/app/src/main/java/com/gsoft/opus/domain/model/Activite.kt
package com.gsoft.opus.domain.model

/**
 * Activite — patrol / intervention record (Service Général).
 *
 * Patrol itineraries live in six nullable fields (one per type×mode pair).
 * A null itinerary means the patrol mode was not selected; an empty string
 * means the mode was selected without a detailed itinerary.
 */
data class Activite(
    val id: Int,
    val dateActivite: String,
    val heureActivite: String,
    val patrouilleDiurneMotoriseeItineraire: String?,
    val patrouilleDiurnePedestreItineraire: String?,
    val patrouilleDiurnePorteeItineraire: String?,
    val patrouilleNocturneMotoriseeItineraire: String?,
    val patrouilleNocturnePedestreItineraire: String?,
    val patrouilleNocturnePorteeItineraire: String?,
    val operationCiblee: String?,
    val faitsConstates: String?,
    val compteRenduHierarchie: String?,
    val conduiteATenir: String?,
    val natureIntervention: String?,
    val suitesDonnees: String?,
    /** GPS latitude captured at record time (mobile only, null on desktop). */
    val latitude: Double?,
    /** GPS longitude captured at record time (mobile only, null on desktop). */
    val longitude: Double?,
    val createdBy: Int?,
    val createdAt: String?,
    val updatedAt: String?,
    val agentUsername: String?,
    val agentPrenoms: String?,
    val agentNom: String?
) {
    /** "HH:MM:SS" from the API is displayed as "HH:MM". */
    val heureDisplay: String get() = heureActivite.take(5)

    /** Display name of the agent who recorded the entry. */
    val agentDisplayName: String
        get() = listOfNotNull(agentPrenoms, agentNom)
            .filter { it.isNotBlank() }
            .joinToString(" ")
            .ifBlank { agentUsername ?: "" }

    /** Itinerary stored for the given patrol type/mode (null = not selected). */
    fun itineraire(type: String, mode: String): String? = when (type to mode) {
        "diurne" to "motorisee" -> patrouilleDiurneMotoriseeItineraire
        "diurne" to "pedestre" -> patrouilleDiurnePedestreItineraire
        "diurne" to "portee" -> patrouilleDiurnePorteeItineraire
        "nocturne" to "motorisee" -> patrouilleNocturneMotoriseeItineraire
        "nocturne" to "pedestre" -> patrouilleNocturnePedestreItineraire
        "nocturne" to "portee" -> patrouilleNocturnePorteeItineraire
        else -> null
    }

    /** Whether at least one patrol mode was selected. */
    val hasPatrouille: Boolean
        get() = listOf(
            patrouilleDiurneMotoriseeItineraire, patrouilleDiurnePedestreItineraire,
            patrouilleDiurnePorteeItineraire, patrouilleNocturneMotoriseeItineraire,
            patrouilleNocturnePedestreItineraire, patrouilleNocturnePorteeItineraire
        ).any { it != null }

    /**
     * Short summary for list rows, e.g. "Diurne motorisée • Nocturne portée".
     * Empty when no patrol mode was selected.
     */
    val patrouilleSummary: String
        get() = PATROUILLE_TYPES.flatMap { (type, typeLabel) ->
            PATROUILLE_MODES.mapNotNull { (mode, modeLabel) ->
                if (itineraire(type, mode) != null) "$typeLabel $modeLabel" else null
            }
        }.joinToString(" • ")

    companion object {
        /** Patrol types in display order: code → French label. */
        val PATROUILLE_TYPES = listOf(
            "diurne" to "Diurne",
            "nocturne" to "Nocturne"
        )

        /** Patrol modes in display order: code → French label. */
        val PATROUILLE_MODES = listOf(
            "motorisee" to "Motorisée",
            "pedestre" to "Pédestre",
            "portee" to "Portée"
        )
    }
}

/** Attached file of an activité (stored on disk server-side). */
data class ActiviteAttachment(
    val id: Int,
    val activiteId: Int,
    val title: String,
    val filename: String,
    val originalFilename: String,
    val mimeType: String?,
    val fileSize: Long?,
    val createdAt: String?
)

File: api/src/Models/Activite.php
<?php
declare(strict_types=1);

namespace App\Models;

use App\Database;

/**
 * Activite — patrol / intervention record (Service Général).
 *
 * Single-table feature: patrol itineraries live in six nullable columns
 * (one per type×mode pair); a NULL itinerary means the mode was not
 * selected, an empty string means it was selected without details.
 */
class Activite
{
    /** The six patrol itinerary columns (NULL = mode not selected). */
    public const ITINERAIRE_FIELDS = [
        'patrouille_diurne_motorisee_itineraire',
        'patrouille_diurne_pedestre_itineraire',
        'patrouille_diurne_portee_itineraire',
        'patrouille_nocturne_motorisee_itineraire',
        'patrouille_nocturne_pedestre_itineraire',
        'patrouille_nocturne_portee_itineraire',
    ];

    /** Other free-text columns (trimmed; empty string → null). */
    public const TEXT_FIELDS = [
        'operation_ciblee', 'faits_constates', 'compte_rendu_hierarchie',
        'conduite_a_tenir', 'nature_intervention', 'suites_donnees',
    ];

    public static function all(array $filters = []): array
    {
        $db = Database::getInstance()->getConnection();
        $where = [];
        $args = [];

        if (!empty($filters['search'])) {
            $where[] = '(a.operation_ciblee LIKE ? OR a.faits_constates LIKE ? OR a.nature_intervention LIKE ? OR a.suites_donnees LIKE ?)';
            $s = '%' . $filters['search'] . '%';
            array_push($args, $s, $s, $s, $s);
        }

        $sql = 'SELECT a.*, u.username AS agent_username, u.personnel_id AS agent_personnel_id, p.firstname AS agent_prenoms, p.lastname AS agent_nom
                FROM activite a
                LEFT JOIN users u ON a.created_by = u.id
                LEFT JOIN personnel p ON u.personnel_id = p.id';
        if ($where) {
            $sql .= ' WHERE ' . implode(' AND ', $where);
        }
        $sql .= ' ORDER BY a.date_activite DESC, a.id DESC';

        $stmt = $db->prepare($sql);
        $stmt->execute($args);
        return $stmt->fetchAll(\PDO::FETCH_ASSOC);
    }

    public static function find(int $id): ?array
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare(
            'SELECT a.*, u.username AS agent_username, u.personnel_id AS agent_personnel_id, p.firstname AS agent_prenoms, p.lastname AS agent_nom
             FROM activite a
             LEFT JOIN users u ON a.created_by = u.id
             LEFT JOIN personnel p ON u.personnel_id = p.id
             WHERE a.id = ?'
        );
        $stmt->execute([$id]);
        $row = $stmt->fetch(\PDO::FETCH_ASSOC);
        return $row !== false ? $row : null;
    }

    public static function create(array $data): int
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare(
            'INSERT INTO activite
                (date_activite, heure_activite,
                 patrouille_diurne_motorisee_itineraire, patrouille_diurne_pedestre_itineraire, patrouille_diurne_portee_itineraire,
                 patrouille_nocturne_motorisee_itineraire, patrouille_nocturne_pedestre_itineraire, patrouille_nocturne_portee_itineraire,
                 operation_ciblee, faits_constates, compte_rendu_hierarchie, conduite_a_tenir,
                 nature_intervention, suites_donnees, latitude, longitude, created_by)
             VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)'
        );
        $stmt->execute([
            $data['date_activite'],
            $data['heure_activite'],
            $data['patrouille_diurne_motorisee_itineraire'] ?? null,
            $data['patrouille_diurne_pedestre_itineraire'] ?? null,
            $data['patrouille_diurne_portee_itineraire'] ?? null,
            $data['patrouille_nocturne_motorisee_itineraire'] ?? null,
            $data['patrouille_nocturne_pedestre_itineraire'] ?? null,
            $data['patrouille_nocturne_portee_itineraire'] ?? null,
            $data['operation_ciblee'] ?? null,
            $data['faits_constates'] ?? null,
            $data['compte_rendu_hierarchie'] ?? null,
            $data['conduite_a_tenir'] ?? null,
            $data['nature_intervention'] ?? null,
            $data['suites_donnees'] ?? null,
            $data['latitude'] ?? null,
            $data['longitude'] ?? null,
            $data['created_by'] ?? null,
        ]);
        return (int) $db->lastInsertId();
    }

    public static function update(int $id, array $data): bool
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare(
            'UPDATE activite SET
                date_activite = ?, heure_activite = ?,
                patrouille_diurne_motorisee_itineraire = ?, patrouille_diurne_pedestre_itineraire = ?, patrouille_diurne_portee_itineraire = ?,
                patrouille_nocturne_motorisee_itineraire = ?, patrouille_nocturne_pedestre_itineraire = ?, patrouille_nocturne_portee_itineraire = ?,
                operation_ciblee = ?, faits_constates = ?, compte_rendu_hierarchie = ?, conduite_a_tenir = ?,
                nature_intervention = ?, suites_donnees = ?, latitude = ?, longitude = ?
             WHERE id = ?'
        );
        return $stmt->execute([
            $data['date_activite'],
            $data['heure_activite'],
            $data['patrouille_diurne_motorisee_itineraire'] ?? null,
            $data['patrouille_diurne_pedestre_itineraire'] ?? null,
            $data['patrouille_diurne_portee_itineraire'] ?? null,
            $data['patrouille_nocturne_motorisee_itineraire'] ?? null,
            $data['patrouille_nocturne_pedestre_itineraire'] ?? null,
            $data['patrouille_nocturne_portee_itineraire'] ?? null,
            $data['operation_ciblee'] ?? null,
            $data['faits_constates'] ?? null,
            $data['compte_rendu_hierarchie'] ?? null,
            $data['conduite_a_tenir'] ?? null,
            $data['nature_intervention'] ?? null,
            $data['suites_donnees'] ?? null,
            $data['latitude'] ?? null,
            $data['longitude'] ?? null,
            $id,
        ]);
    }

    public static function delete(int $id): bool
    {
        $db = Database::getInstance()->getConnection();
        $stmt = $db->prepare('DELETE FROM activite WHERE id = ?');
        return $stmt->execute([$id]);
    }
}
